server/src/main/java/org/elasticsearch/reservedstate/service/ReservedStateUpdateTask.java

```diff
@@ -185,6 +189,14 @@ static boolean checkMetadataVersion(
         }
 
         if (existingMetadata != null && existingMetadata.version() >= reservedStateVersion.version()) {
+            if (allowSameVersion && existingMetadata.version().equals(reservedStateVersion.version())) {
```

Comment:
Passing through this "allowSameVersion" is quite invasive on the API. It feels clunky, especially seeing the use here, first checking for `>=` then checking for `=`.

Instead, I wonder if we could adjust the passed in `reservedStateVersion`? So view it as a minimum version for processing. Most of the time we would take the existing version and add 1. In the case of reprocessing we just take the existing version. Wdyt?

Reply:
@rjernst you're saying adjust `reservedStateVersion` here https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/reservedstate/service/ReservedClusterStateService.java#L177? 

That would still leave quite a few methods with the boolean flag since this method gets called by a method that gets passed an `XContent` parser.

If we adjust the version, we'd also need to account for special version sentinels e.g.:

```
reservedStateVersion.version().equals(ReservedStateMetadata.EMPTY_VERSION)
``` 

Likewise, I think the logging will be more confusing since there'd be versions that don't match actual data.  

I think adjusting the version would functionally work if we accommodate for the special sentinel versions, but would make the code harder to reason about and easier to get wrong. Also, I have it in the back of my mind that ECK uses timestamps for the version field, which I _think_ is fine but I'd need to think about it more. 

To avoid some of the signature changes WDYT about adding the boolean flag to `ReservedStateChunk` instead? 

> especially seeing the use here, first checking for >= then checking for =.
  
I can for sure refactor that to be less of an eye sore (I know that's not your main objection, but just as a note).

Reply:
This is roughly what moving the boolean flag to `ReservedStateChunk` would look like: https://github.com/n1v0lg/elasticsearch/commit/ae00eaa48e1485fdba9a8b81a3da7a1619b890f6

I'm not strongly opposed to adjusting the version instead but do worry that it might be harder to reason about. LMKWYT

Reply:
Discussed this on Slack yesterday: `reservedStateVersion` is the version read from file-settings, not the current version stored in metadata. We'd want to modify the current version stored in metadata (instead of `reservedStateVersion`)  but this is challenging since we read the metadata version from cluster state, twice during processing, without an immediately obvious way of marking it as requiring or not requiring an adjustment. 

I've provisionally pushed an alternative refactor that moves the stray boolean flag into an object held by `ReservedStateChunk` and cleaned up `checkMetadataVersion`. 

I'll give adjusting versions as an alternative more thought but wasn't able to come up with a clean way to use that approach.

Reply:
As discuss on Slack: instead of passing a boolean flag, going with a `BiPredicate` for the comparison instead.